### ISSUE TITLE
[v41] Too high CPU usage in some cases if vehicles used by bots.

### ISSUE THREAD

**Issue by an-eternity** [SeriousBuggie/XVehicles#3]
This issue is difficult to reproduce. It happens only when bot is in State `Wandering` or `Roaming` with `Enemy==None`, and only in some maps...
Also, it seems, CPU usage in this case depends on something related to the size of the area around... In a large area CPU usage increase produced by a single vehicle instance may reach up to 10 times.

Way to reproduce:
1. Open map MH-EnterToCore-Part3.
2. Set `CreatureFactory0` State to `Finished`.
3. Set your player location to `Teleporter0`.
4. Spawn bots there as that location.
5. Spawn several vehicles there at that location (better - stationary of flying vehicles, because other vehicles will fall to damage zone...)
6. When at least 1 bot begins to use vehicle, CPU usage gets increased at least 10 times.
7. Then, if to enter console command `killall scriptedpawn`, issue disappears and CPU usage returns to normal.

**Comment by an-eternity** [comment]
Tested v44 just now - still can reproduce this issue the way described above.
Maybe this problem is not in this mod itself?

**Comment by SeriousBuggie** [comment]
Still not able reproduce.
Here test map.
[MH-EnterToCore-Part3_bug3.zip](https://github.com/SeriousBuggie/XVehicles/files/9333967/MH-EnterToCore-Part3_bug3.zip)
I start it as spectator inn practice session. look into wall (for reduce rendering CPU load), add 1 bot - no Any CPU change. Even when bot jump into manta. I kill all scriptedpawn - no any CPU usage change.

**Comment by SeriousBuggie** [comment]
Issue still in place?

**Comment by an-eternity** [comment]
In the last release i tested (v48) this issue was still in place.
Will try to test v51 today, then will reply results.

I have inspected the scripts, but can't track down an exact place where this performance loss occurs. I suppose, it calls Collision Hash slow native functions way too often.
If [this](https://github.com/OldUnreal/UnrealTournamentPatches/issues/72) and some else debugging options (functions execution time stats, etc.) were available, maybe it was not that hard to figure it out.

**Comment by an-eternity** [comment]
> Here test map. [MH-EnterToCore-Part3_bug3.zip](https://github.com/SeriousBuggie/XVehicles/files/9333967/MH-EnterToCore-Part3_bug3.zip)

By the way, i noticed some problem (it was also related to my test server config) when the count of bots or players that are about to spawn simultaneously is more than the count of spawn points available out there:
_"Critical: MonsterHunt MH-EnterToCore-Part3_bug2.MonsterHunt0 (Function Botpack.TeamGamePlus.FindPlayerStart:0000) Infinite script recursion (250 calls) detected"_
Also, depending on server settings function `DeathMatchPlus.SpawnBot()` can produce script error in line `((bHumansOnly || Level.bHumansOnly) && !NewBot.bIsHuman)` if bot is failed to spawn.

**Comment by SeriousBuggie** [comment]
Last one not related to XV or this issue. So better report it on OU github instead. Possible as two separate issues.

**Comment by an-eternity** [comment]
> Issue still in place?

Yes.
Tested v51 now - result is the same as with previous versions.

**Comment by an-eternity** [comment]
If i understand UTrace logs properly, it shows `Bot.AttitudeTo()` -> `AssessBotAttitude()` is being called (for every ScriptedPawn) from `WeaponAttachment.FindEnemy()` very often (in every Tick event).
Test server i use has modified `AssessBotAttitude()` function, it calls slow native function `TraceActors()` to make bot ignore targets located behind the movers.
This explains why CPU load in this case is in correlation with the map size, scripted pawns and bots count.
It seems, i have to add a distance limit, so that to not call `TraceActors()` for every ScriptedPawn on the map which is not reachable due to distance anyway.

Normally, bots don't iterate an entire PawnList on every tick, but still are able to find the target quickly enough.
This means, issue consists of 2 parts which produce high CPU load if combined together.

I think it would be better to rely on the standard Bot behavior/logic in this case, not override it and not force custom targets. Or at least to provide an ini option that allows to select a desired behavior.

I will fix my `AssessBotAttitude()` function and then will reply a results...

**Comment by SeriousBuggie** [comment]
You must always guard TraceActors and Trace with FastTrace first. It named Fast by reason. So this can cut most huge calculations.

> I think it would be better to rely on the standard Bot behavior/logic in this case, not override it and not force custom targets. Or at least to provide an ini option that allows to select a desired behavior.

Not agree. This code appear by reason. This is reason - high range + high fire power of vehicles. Stock sight very limited. bots mostly blind. So they never shoot far away when player does. So this lead to simple use bots as target practice when stay far enough. I able simple win on bloodgulch stay far away and fire from red tank. This enough for suppress any defense and allow bots from my team bring flag.

In general if you alter some stuck functions you need be very careful. Code usually tested with stock implementation and can call some function thousands per seconds. If you introduce here high load calculation it is completely your responsibility ensure it kill everything. Hardly expect low usage of low costs function.

**Comment by an-eternity** [comment]
`WeaponAttachment` has almost equal function - `TraceHit()`, that is called if condition `Bot.AttitudeTo(P) > ATTITUDE_Frenzy` returns `False`. The only significant difference is that `AssessBotAttitude()` on my side doesn't have 'FastTrace()' guard prior to `TraceActors()`. I will add `FastTrace()` then.

Is there a max limit of distance to target for Bot in the vehicle, if yes then what is the value?

**Comment by SeriousBuggie** [comment]
Yes. And I already experience issues with it, so add FastTrace here. :)

Vehicle able shoot in distance for near 80k. So this mean it can effectively fire to any place on map.
So no. There no any such limit.

**Comment by an-eternity** [comment]
I see v41 didn't have `TraceHit()` function at all. This explains why this issue was reproduceable with v41 when i tested it with default server installation without mods. It was half a year ago back then, i considered problem was isolated to either XV mod or to standard UC/native part, and then and so far was testing it only with my custom test server setup.
Also, v41 didn't call to `Bot.AttitudeTo(P)` on PawnList iteration. Then later you have added `FastTrace()`, and also added `Bot.AttitudeTo(P)` which follows to my slow version of `TraceHit()` that didn't use `FastTrace()` anyway, so test results i was getting didn't change.

PawnList iteration script used in `WeaponAttachment.FindEnemy()` in v41:
```
	For( P=Level.PawnList; P!=None; P=P.NextPawn )
	{
		if (P.Health <= 0 || (P.PlayerReplicationInfo != None && 
			(P.PlayerReplicationInfo.bIsSpectator || P.PlayerReplicationInfo.Team == OwnerVehicle.CurrentTeam)))
			continue; // check not spectator and other stuff
//		if (!Bot.LineOfSightTo(P)) continue;
		HL = P.Location;
		if (PitchPart != None)
			HN = PitchPart.Location;
		else
			HN = Location;
		Hit = ViewActor.Trace(HL, HN, HL, HN, true);
		if (Hit != P && (DriverWeapon(P.Weapon) == None || DriverWeapon(P.Weapon).VehicleOwner != Hit))
			continue;
		HL = P.Location - OwnerVehicle.Location;
		// X dot X == VSize(X)*VSize(X)
		Dist = HL dot HL;
		if (Best == None || Dist < BestDist)
		{
			Best = P;
			BestDist = Dist;
		}
	}
```

I will repeat the test with fixed `AssessBotAttitude()` today as soon as will be on the PC...

> Vehicle able shoot in distance for near 80k.

I think to add a distance check prior to `FastTrace()`, maybe it makes sense in case of very huge map, `VSize()` must work faster...

**Comment by SeriousBuggie** [comment]
Yes. As I mention it be wrong. v41 implementation is bad. And it be fixed. `AttitudeTo` not count, used high-cost call and this make high CPU load and force attack allies, like Nali or Cows.

On net play splash damage penetrate movers. Most XV weapons produce it. So it little cheatish, but bot can hit you behind closed mover.

**Comment by an-eternity** [comment]
Tested... With fixed `AssessBotAttitude()` this issue doesn't happen.
In the test case i always used for this issue so far (T/R == 35, bots count == 3, scripted pawns count == 156) this function was called about 34700 times per second, which caused core overload to any CPU i have tested it with (even those with fastest single core performance).
Now, with `TraceActors()` guarded by `FastTrace()`, no any significant increase in CPU load takes place, even if `AssessBotAttitude()` is being called 100000 times per second.
With no XVehicles in this test case, 3 bots call this function only about 25-30 times per second.

Also, significant impact on the test results was produced by the map geometry and the location chosen to place bots with XVehicles in - trace function must check quite a long distance from there (over than 10k units for every scripted pawn) until hit Level.

I have added a distance check (80000 units limit) prior to `FastTrace()` call, but in this test case it makes no sense because the most distant scripted pawn is located at 48000 units away from the bots.

Custom `AssessBotAttitude()` implementation i use for tests also takes care of all non-enemy pawns, including known non-enemy Nali and Cow classes.
And this function seems to be the only place where trace checks can prevent bots from taking targets behind the movers. 

> On net play splash damage penetrate movers. Most XV weapons produce it. So it little cheatish, but bot can hit you behind closed mover.

Even some standard projectiles can do this...
Originally an idea to trace the line to target in `AssessBotAttitude()` was suggested by Sektor and Higor [here](https://ut99.org/viewtopic.php?f=64&p=70136#p70136).
Now, i think of improved version of it, which randomly allows bot to take the target behind the mover if that target is located close enough to the mover, but it is complicated because need to make sure whether bot uses an appropriate weapon which is able to reach that target behind the mover... Anyway, it is already unrelated to this issue.

I think everything is clear now, no questions left about this issue, so it can be closed now.

Thanks for info about UTrace, it is very powerful tool for debugging, it helped to track down this problem within a few minutes today, so this long-lasting issue is resolved finally. No idea, how more long it would have taken without UTrace, maybe never....
